### PR TITLE
chore(codecs): Implement common framing and codec configs for sources via `tokio_util::codec::Decoder`

### LINKED ISSUE #7463: Implement common `decoding` config for sources
Per Framing and Codecs RFC: #8620

Add `decoding` config that can be shared by all relevant sources.

### PR DESCRIPTION
**Pull request by pablosichert**
Closes #7463.

(For context: Previous discarded approaches: #7482, #8002.)

Todos:
- [x] Add documentation for new traits / `codec` module using `#![deny(missing_docs)]`
- [x] Add documentation for `codec` implementations using `#![deny(missing_docs)]`
- [x] Provide common configuration
- [x] Add cue docs
- [x] Reuse `Decoder` for sources that internally build an `Event` from bytes
    - [x] `fluent` source
    - [x] `logstash` source
    - [x] `statsd` source
    - [x] `syslog` source
    - [x] `vector` source
- [x] Integrate `DecodingConfig` into sources that can expose framing/decoding to users:
    - [x] `kafka` source
    - [x] `socket` source
        - [x] TCP
        - [x] UDP
        - [x] Unix
    - [x] `stdin` source
    - [x] `http` source

Additional candidates for future PRs:
- `file` source
- `aws_kinesis_firehose` source
- `aws_s3` source

### CODE REVIEW

File: src/sources/util/decoding/mod.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+impl From<LinesCodecError> for Error {
+    fn from(error: LinesCodecError) -> Self {
+        Self {
+            error: std::sync::Arc::new(error),
+        }
+    }
+}
+
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes, Error = Error> + Send + Sync>,
```
Comment (lukesteensen) [inline src/sources/util/decoding/mod.rs:58]: We're repeating this type a lot and it's pretty complex. Should we define our own trait like we did for `DecoderError`?

File: src/sources/util/decoding/mod.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+impl<Error: From<std::io::Error> + Into<self::Error>, Item: Into<Bytes>> tokio_util::codec::Decoder
+    for BytesDecoder<Error, Item>
+{
+    type Item = Bytes;
+    type Error = self::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.decoder
+            .decode(buf)
+            .map(|result| result.map(|frame| frame.into()))
```
Comment (lukesteensen) [inline src/sources/util/decoding/mod.rs:133]: I'm a little bit concerned that this whole `BytesDecoder` wrapper thing will obfuscate the `Into<Bytes>` operation, which could involve undesirable and performance-impacting copying. We should try to work directly with `Bytes` where possible, but it should be explicitly handled when there's a source or decoder that doesn't.

File: src/sources/util/tcp.rs [hunk 3]
```diff
@@ -53,41 +54,27 @@ async fn make_listener(
...
-    fn build_ack(&self, _frame: &<Self::Decoder as Decoder>::Item) -> Bytes {
+    fn handle_event(&self, event: &mut Event, host: Bytes, byte_size: usize);
+
+    fn build_ack(&self, _item: &Self::Item) -> Bytes {
         Bytes::new()
     }
 
     fn run(
-        self,
+        self: Arc<Self>,
```
Comment (lukesteensen) [inline src/sources/util/tcp.rs:77]: A comment on why we need `Arc<Self>` would be nice. Unless I'm missing something, it's not obvious from the implementation.

File: src/sources/util/mod.rs [hunk 4]
```diff
@@ -1,5 +1,14 @@
...
 mod body_decoding;
+#[cfg(any(
+    feature = "sources-kafka",
+    feature = "sources-socket",
+    feature = "sources-statsd",
+    feature = "sources-stdin",
+    feature = "sources-syslog",
+    feature = "sources-vector",
+))]
+pub mod decoding;
```
Comment (lukesteensen) [inline src/sources/util/mod.rs:11]: I'm really not sure of the value of doing all this feature gating and nesting within `sources/util/`. If you asked me, I'd move all of this to a top-level `codecs/` and not worry about feature flags unless something brings in relatively heavy dependencies. That said, it's not something I feel strongly about.

**Review by lukesteensen — COMMENTED** [review]
This is looking solid so far! Dropping a few notes now but will continue working through it in more detail.

File: src/sources/util/mod.rs (hunk 4 above)
Comment (pablosichert) [reply src/sources/util/mod.rs:11]: Just tried moving everything from `src/sources/util/decoding` to `lib/codec`. However, it's not easily possible since we depend on `src/internal_events` and other utils from `src/sources/util`

File: src/sources/util/decoding/mod.rs (hunk 2 above)
Comment (pablosichert) [reply src/sources/util/decoding/mod.rs:133]: `BytesCodec` exists to reuse codecs where the output is not `Item = Bytes, Error = decoding::Error`, but can be converted trivially.

Such as `LengthDelimitedCodec` (which has `Item = BytesMut, Error = io::Error`), `LinesCodec` (which has `Item = String, Error = LinesCodecError`) and `OctetCountingDecoder` (which has `Output = Bytes, Error = LinesCodecError`).

The conversion between `BytesMut` or `String` to `Bytes` is zero-copy, and as far as I can see `Bytes` doesn't implement the `From` trait when it wouldn't be zero-copy. However, that wouldn't prevent us from implementing a type that implements the `From` trait for `Bytes` in a possibly expensive fashion.

> We should try to work directly with Bytes where possible, but it should be explicitly handled when there's a source or decoder that doesn't.

How would you approach this - would you like to define new types e.g. `LengthDelimitedBytesCodec` for `LengthDelimitedCodec` that internally goes from `String` to `Bytes`? I'm not sure if that would obfuscate the conversion less since you still need to know if the `into` implementation is zero-copy

File: src/sources/util/mod.rs (hunk 4 above)
Comment (pablosichert) [reply src/sources/util/mod.rs:11]: Unfortunately, as long as we can't extract those pieces that we depend on to another library, the feature gating is needed to prevent clippy from reporting unused code

File: src/sources/fluent.rs [hunk 5]
```diff
@@ -226,19 +244,19 @@ impl FluentDecoder {
...
 impl Decoder for FluentDecoder {
-    type Item = FluentFrame;
+    type Item = (FluentFrame, usize);
     type Error = DecodeError;
 
     fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        if let Some(frame) = self.unread_frames.pop_front() {
-            return Ok(Some(frame));
+        if let Some((frame, byte_size)) = self.unread_frames.pop_front() {
+            return Ok(Some((frame, byte_size)));
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/sources/fluent.rs]: Adressed in 7005cdafebfcd543a2151b8ae6aa0f64ec1e1521

File: src/sources/util/decoding/config.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        let framer: BoxedFramer = match self.framing {
+            Some(framing) => framing.build(),
+            None => Box::new(super::BytesDecoder::new(BytesDelimitedCodec::new(b'\n'))),
+        };
+
+        let parser: BoxedParser = match self.decoding {
+            Some(ParserConfig::Bytes) | None => Box::new(super::BytesParser),
+            #[cfg(feature = "sources-syslog")]
+            Some(ParserConfig::Syslog) => Box::new(super::SyslogParser),
+        };
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/sources/util/decoding/config.rs]: Good suggestion, and also thanks to your comment on the RFC, both framing config and parser config expose a build method and are dynamically selected via `typetag`

File: src/sources/util/tcp.rs (hunk 3 above)
Comment (pablosichert) [reply src/sources/util/tcp.rs]: This is a good question, and the reason for this has been hidden quite a few layers deep.

For the `TcpSource` to be `Clone`, `Decoder` needs to be clone. However, it stores `Box`es of a `Framer` and `Parser`. `Box` is not `Clone`, but can be made via `dyn_clone::DynClone`. However, that doesn't work for structs that store a `Box` with a generic type associated to the struct, and previously the `BytesDecoder` wrapper stored a generic type that was `Into<Bytes>`.

Using `Arc` instead of `Box` there also doesn't work there, because the inner `Decoder` needs to be mutated via `decode` calls and `Decoder` also needs to be `Send + Sync` because `TcpSource` can spawn multiple instances concurrently (one per request). I felt that adding `Arc<Mutex<_>>` was too much cost for an operation that is executed for every frame.

Therefore, I decided to use `Arc` for cloning the source and build a fresh decoder using `source.build_decoder` for each request.

---

Long story short: I went back to forcing `Decoder` to be `Clone` in 655ac4e64c79de39f6567bfe8a60f8c6fb2e0bb4 by identifying how to accomplish that. It required getting rid of `BytesDecoder` (which @lukesteensen also suggested due to implicit conversion into `Byte`), making the `Box`es `Clone` by using `dyn_clone::DynClone` and finally hacking in a `Clone` implementation for `tokio_util::codec::LengthDelimitedCodec` that builds it anew instead of cloning its internal state, because unfortunately it does not implement `Clone` even though all its internal fields do, see https://github.com/timberio/vector/blob/655ac4e64c79de39f6567bfe8a60f8c6fb2e0bb4/src/sources/util/decoding/framing/length_delimited.rs#L33-L40.

File: src/sources/util/decoding/mod.rs (hunk 2 above)
Comment (pablosichert) [reply src/sources/util/decoding/mod.rs]: I removed `BytesCodec` and specified the `Item` type for all our decoders explicitly, also see https://github.com/timberio/vector/pull/8236#discussion_r689127860 for some more context

File: src/sources/util/decoding/mod.rs (hunk 1 above)
Comment (pablosichert) [reply src/sources/util/decoding/mod.rs]: This has been updated accordingly:

https://github.com/timberio/vector/blob/e1065d4e30975db3ff31cd93e372a1d19d1d8070/src/sources/util/decoding/mod.rs#L90-L98

File: src/sources/util/decoding/mod.rs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+pub trait Parser {
+    fn parse(&self, bytes: Bytes) -> crate::Result<Event>;
+}
+
+pub trait DecoderError: std::error::Error + TcpError + Send + Sync {}
+
+#[derive(Debug)]
+pub struct Error {
+    error: Arc<dyn DecoderError>,
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/sources/util/decoding/mod.rs]: This has also been victim to https://github.com/timberio/vector/pull/8236#discussion_r689127860, but was not required after the mentioned refactoring any more

File: src/sources/util/decoding/config.rs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        Self { framing, decoding }
+    }
+}
+
+impl DecodingConfig {
+    pub fn build(&self) -> Decoder {
+        let framer: BoxedFramer = match self.framing {
+            Some(framing) => framing.build(),
+            None => Box::new(super::BytesDecoder::new(BytesDelimitedCodec::new(b'\n'))),
+        };
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/sources/util/decoding/config.rs]: Looks like it doesn't work to implicitly convert between `&T` and `Box<T>` in `map`:

```
type mismatch in function arguments
expected signature of `fn(Box<dyn FramingConfig>) -> _`rustcE0631
config.rs(8, 5): found signature of `for<'r> fn(&'r _) -> _`
```

File: src/sources/util/mod.rs (hunk 4 above)
Comment (pablosichert) [reply src/sources/util/mod.rs]: Moved everything to top-level `codec` in ce1a6d35ac2d3fa8493700424e8e910d04285e65

File: src/sources/util/decoding/config.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sources/util/decoding/config.rs]: Ah, I missed that `framing` would be a `Box`. The long form would be `.map(|framing| framing.build())`

**Comment by pablosichert** [comment]
All of the changes in this PR have been incorporated by the referenced PRs.
